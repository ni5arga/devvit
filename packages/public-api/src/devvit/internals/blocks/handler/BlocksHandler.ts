import {
  EffectType,
  type Effect,
  type Metadata,
  type UIEvent,
  type UIRequest,
  type UIResponse,
} from '@devvit/protos';
import { ContextBuilder } from './ContextBuilder.js';
import { BlocksTransformer } from '../BlocksTransformer.js';
import type { BlockElement } from '../../../Devvit.js';
import type { ReifiedBlockElement, ReifiedBlockElementOrLiteral } from '../BlocksReconciler.js';
import type { Hook, HookSegment, HookParams, Props, BlocksState } from './types.js';
import { RenderInterruptError } from './types.js';
import { RenderContext } from './RenderContext.js';
import type { JSONValue } from '@devvit/shared-types/json.js';
import type { EffectEmitter } from '../EffectEmitter.js';

/**
 * This can be a global/singleton because render is synchronous.
 *
 * If you want to use this from somewhere else, please consider using one of the
 * functions like isRendering or registerHook, and then try to add additional
 * functions here if needed.  Don't use this directly.
 */
export let _activeRenderContext: RenderContext | null = null;

export function useEffectEmitter(): EffectEmitter {
  if (!_activeRenderContext) {
    throw new Error('Hooks can only be declared at the top of a component.');
  }
  return _activeRenderContext;
}

export function isRendering(): boolean {
  return _activeRenderContext !== null;
}

function _structuredClone<T extends JSONValue>(obj: T): T {
  return JSON.parse(JSON.stringify(obj));
}

/**
 * This is the recommended low-level interface for creating hooks like useState or useAsync.
 *
 * Practically, this initializes your hook if it doesn't already exist, and makes sure
 * that its state gets all sync'd up.
 *
 * @param HookSegment -- A name for this hook.  This is used to dedupe hooks.
 * @param initializer
 *    factory for building this hook
 * @returns
 */
export function registerHook<H extends Hook>(
  options: HookSegment,
  initializer: (p: HookParams) => H
): H {
  if (!_activeRenderContext) {
    throw new Error(
      "Hooks can only be declared at the top of a component.  You cannot declare hooks outside\
    of components or inside of event handlers.  It's almost always a mistake to declare hooks inside of loops or\
    conditionals."
    );
  }
  const hookId = _activeRenderContext.nextHookId(options);
  const context = _activeRenderContext;
  const params: HookParams = {
    hookId,
    changed: () => {
      context._changed[hookId] = true;
      context._state[hookId] = context?._hooks[hookId]?.state;
    },
    context: _activeRenderContext,
  };
  const fromNull = _activeRenderContext._state[hookId] === undefined;
  _activeRenderContext._hooks[hookId] = _activeRenderContext._hooks[hookId] ?? initializer(params);
  const hook: H = _activeRenderContext._hooks[hookId] as H;

  if (_activeRenderContext._state[hookId] !== undefined) {
    hook.state = _activeRenderContext._state[hookId];
  }
  if (hook.onLoad) {
    hook.onLoad(_activeRenderContext);
  }
  if (fromNull && hook.state !== undefined && hook.state !== null) {
    params.changed();
  }
  return hook;
}

export let _latestBlocksHandler: BlocksHandler | null = null;

/**
 * Replacing BlocksReconciler, the model is now less of a "reconciliation", and more
 * of a handling a request/response lifecycle.
 *
 */
export class BlocksHandler {
  #root: JSX.ComponentFunction;
  #contextBuilder: ContextBuilder = new ContextBuilder();
  #blocksTransformer: BlocksTransformer = new BlocksTransformer();
  _latestRenderContext: RenderContext | null = null;

  constructor(root: JSX.ComponentFunction) {
    this.#root = root;
    _latestBlocksHandler = this;
  }

  async handle(request: UIRequest, metadata?: Metadata): Promise<UIResponse> {
    const context = new RenderContext(request);
    const devvitContext = this.#contextBuilder.buildContext(context, metadata!);
    context.devvitContext = devvitContext;

    let blocks;

    /**
     * Events on the main queue must be handled in order, so that state is updated in the correct order.  Events
     * on other queues can be handled in parallel, because they only emit effects.
     *
     * There is an optimization here to process SendEventEffects locally, instead of letting them bubble up to the
     * platform.  This prevents a round trip to the platform for every event.
     *
     * This also means we need to respect execution queues here, and not just in the platform.
     */
    const eventsToProcess = request.events;
    const noEvents = !request.events?.length;
    const isMainQueue = noEvents || eventsToProcess.some((e) => !e.async);

    const isBlockingSSR = eventsToProcess.some((e) => e.blocking);

    let progress:
      | {
          _state: BlocksState;
          _effects: { [key: string]: Effect };
        }
      | undefined;
    let remaining: UIEvent[] = [...eventsToProcess];

    while (eventsToProcess.length > 0) {
      /**
       * A concurrently executable batch is a set of events that can be executed in parallel.  This either one main queue event,
       * or any number of other queue events.
       */
      const batch = [];
      if (!eventsToProcess[0].async) {
        batch.push(eventsToProcess.shift()!);
      } else {
        while (eventsToProcess[0]?.async) {
          batch.push(eventsToProcess.shift()!);
        }
      }
      if (!batch.length) throw Error('batch must have at least one event');
      try {
        if (batch[0].async) {
          const stateCopy = _structuredClone(context._state);
          await this.#handleAsyncQueues(context, ...batch);
          // enforce that state updates are only allowed on the main queue.
          context._state = stateCopy;
        } else {
          await this.#handleMainQueue(context, ...batch);
        }
      } catch (e) {
        /**
         * If we have a progress, we can recover from an error by rolling back to the last progress, and then letting the
         * remaining events be reprocessed.
         */
        if (progress) {
          context._state = progress._state;
          context._effects = progress._effects;
          remaining.forEach((e, i) => {
            const effect: Effect = {
              type: EffectType.EFFECT_SEND_EVENT,
              sendEvent: {
                event: e,
                jumpsQueue: true,
              },
            };
            context.emitEffect(`remaining-${i}`, effect);
          });
          break;
        } else {
          throw e;
        }
      }

      /**
       * If we have any SendEventEffects, we can push them back on the queue to process them locally
       */
      for (const [key, effect] of Object.entries(context._effects)) {
        if (effect.sendEvent?.event) {
          if (!isMainQueue && !effect.sendEvent?.event?.async) {
            // We're async, this is a main queue event.  We need to send it back to the platform to let
            // the platform synchronize it.
            break;
          }

          if (isMainQueue && effect.sendEvent?.event?.async && !isBlockingSSR) {
            // We're main queue, and this is an async event.  We're not in SSR mode, so let's prioritize
            // returning control quickly to the platform so we don't block event loops.
            break;
          }

          //Ok, we can handle this event locally.
          const event = effect.sendEvent.event;
          eventsToProcess.push(event);
          delete context._effects[key];
        }
      }

      /**
       * If we're going back through this again, we need to capture the progress, and the remaining events.
       */
      if (eventsToProcess.length > 0) {
        progress = {
          _state: _structuredClone(context._state),
          _effects: { ...context._effects },
        };
        remaining = [...eventsToProcess];
      }
    } // End of while loop

    if (isMainQueue) {
      // Rendering only happens on the main queue.
      const tags = this.#renderRoot(this.#root, context._rootProps ?? {}, context);
      if (tags) {
        blocks = await this.#blocksTransformer.createBlocksElementOrThrow(tags);
        blocks = await this.#blocksTransformer.ensureRootBlock(blocks);
      }
    }

    return {
      state: context._state,
      effects: context.effects,
      blocks,
    };
  }

  #loadHooks(context: RenderContext, ..._events: UIEvent[]): void {
    // TBD: partial rendering
    this.#renderRoot(this.#root, context.request.props ?? {}, context);
  }

  /**
   * These can all run in parallel, because they only emit effects
   */
  async #handleAsyncQueues(context: RenderContext, ...batch: UIEvent[]): Promise<void> {
    this.#loadHooks(context, ...batch);

    await Promise.all(
      batch.map(async (event) => {
        if (!event.async) {
          throw new Error(
            "You can't mix main and other queues in one batch.  This is likely a platform bug.  Please file an issue in the Discord for someone to help! https://discord.com/channels/1050224141732687912/1115441897079574620"
          );
        }
        await this.#attemptHook(context, event);
      })
    );
  }

  async #attemptHook(context: RenderContext, event: UIEvent): Promise<void> {
    const hook = context._hooks[event.hook!];
    if (hook && hook.onUIEvent) {
      try {
        await hook.onUIEvent(event, context);
      } catch (e) {
        console.error('Error in event handler', e);
        throw e;
      }
    } else {
      await context.handleUndeliveredEvent(event);
    }
  }

  async #handleMainQueue(context: RenderContext, ...batch: UIEvent[]): Promise<void> {
    // We need to handle events in order, so that the state is updated in the correct order.
    for (const event of batch) {
      this.#loadHooks(context, event);
      context._state.__generation = Number(context._state.__generation ?? 0) + 1;
      await this.#attemptHook(context, event);
    }
  }

  #renderRoot(
    component: JSX.ComponentFunction,
    props: Props,
    context: RenderContext
  ): ReifiedBlockElement | undefined {
    context._generated = {};
    _activeRenderContext = context;
    this._latestRenderContext = context;
    try {
      const roots = this.#render(component, props, context);
      if (roots.length !== 1) {
        throw new Error('only one root');
      }
      const root = roots[0];
      if (typeof root === 'string') {
        throw new Error(
          'There must be a root tag.  Try wrapping your app in a <text></text>, <vstack> or other tag.'
        );
      }
      return root;
    } catch (e) {
      if (e instanceof RenderInterruptError) {
        return undefined;
      } else {
        throw e;
      }
    } finally {
      _activeRenderContext = null;
    }
  }

  #render(
    component: JSX.ComponentFunction,
    props: Props,
    context: RenderContext
  ): ReifiedBlockElementOrLiteral[] {
    context.push({ namespace: component.name, ...props });
    try {
      const element = component(props, context.devvitContext);
      return this.#renderElement(element, context);
    } finally {
      context.pop();
    }
  }

  #renderList(list: JSX.Element[], context: RenderContext): ReifiedBlockElementOrLiteral[] {
    list = list.flat(Infinity);
    return list.flatMap((e, i) => {
      if (e && typeof e === 'object' && 'props' in e) {
        if (!e.props?.key) {
          e.props = e.props ?? {};
          e.props.key = i;
        }
      }
      return this.#renderElement(e, context);
    });
  }

  #renderElement(element: JSX.Element, context: RenderContext): ReifiedBlockElementOrLiteral[] {
    if (Array.isArray(element)) {
      return this.#renderList(element, context);
    } else if (isBlockElement(element)) {
      if (element.type === undefined) {
        return this.#renderList(element.children, context);
      } else if (typeof element.type === 'function') {
        const propsWithChildren = { ...element.props, children: element.children };
        return this.#render(element.type, propsWithChildren, context);
      } else {
        context.push({ namespace: element.type, ...element.props });
        const reifiedChildren = this.#renderList(element.children, context);
        const reifiedProps = this.#reifyProps(element.props ?? {});
        context.pop();
        return [{ type: element.type, children: reifiedChildren, props: reifiedProps }];
      }
    } else {
      return [(element ?? '').toString()];
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  #reifyProps(props: { [key: string]: any }): { [key: string]: string } {
    const reifiedProps: { [key: string]: string } = {};
    for (const key in props) {
      if (typeof props[key] === 'function') {
        const hook = registerHook(
          {
            namespace: key,
            key: false,
          },
          ({ hookId }) => ({ hookId, state: null, onUIEvent: props[key] })
        );
        reifiedProps[key] = hook.hookId;
        if ('captureHookRef' in props[key]) {
          props[key].captureHookRef();
        }
      } else {
        const value = props[key];
        if (value !== undefined && value !== null) {
          reifiedProps[key] = value.toString();
        }
      }
    }
    return reifiedProps;
  }
}

function isBlockElement(e: JSX.Element): e is BlockElement {
  return typeof e === 'object' && e != null && 'type' in e;
}
