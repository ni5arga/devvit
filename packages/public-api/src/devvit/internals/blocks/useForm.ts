import { getFormValues } from '../../../apis/ui/helpers/getFormValues.js';
import type { Form, FormFunction, FormKey, FormValues } from '../../../types/form.js';
import type { UseFormHook, UseFormHookState } from '../../../types/hooks.js';
import { Hook } from '../../../types/hooks.js';
import type { BlocksReconciler } from './BlocksReconciler.js';

export function makeUseFormHook(reconciler: BlocksReconciler): UseFormHook {
  function useForm(
    form: Form | FormFunction,
    onSubmit: (values: FormValues) => void | Promise<void>
  ): FormKey {
    const hookIndex = reconciler.currentHookIndex;
    const componentKey = reconciler.getCurrentComponentKey();
    const currentState = reconciler.getCurrentComponentState<UseFormHookState>();
    const previousState = reconciler.getPreviousComponentState<UseFormHookState>();

    const formKey: FormKey = `form.hook.${componentKey}.${hookIndex}`;

    let hookState: UseFormHookState = {
      formKey,
      preventSubmit: false,
      type: Hook.FORM,
    };

    if (hookIndex in currentState) {
      hookState = currentState[hookIndex];
    } else if (hookIndex in previousState) {
      hookState = previousState[hookIndex];
    }

    currentState[hookIndex] = hookState;
    reconciler.forms.set(formKey, form);

    const formSubmittedEvent = reconciler.formSubmittedEvent;

    if (formSubmittedEvent && !hookState.preventSubmit) {
      if (formSubmittedEvent.formId === formKey) {
        reconciler.runHook(async () => {
          const response = onSubmit(getFormValues(formSubmittedEvent.results));

          if (response && response instanceof Promise) {
            await response;
          }

          reconciler.rerenderIn(0);
        });
      }
    }

    currentState[hookIndex].preventSubmit = false;
    reconciler.currentHookIndex++;

    return formKey;
  }

  return useForm;
}
