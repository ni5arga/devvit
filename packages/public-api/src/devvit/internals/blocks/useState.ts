import type { SetStateAction, UseStateHook, UseStateResult } from '../../../types/hooks.js';
import type { BlocksReconciler } from './BlocksReconciler.js';

export function makeUseStateHook(reconciler: BlocksReconciler): UseStateHook {
  function useState<S>(initialState: S): UseStateResult<S> {
    const hookIndex = reconciler.currentHookIndex;
    const currentState = reconciler.getCurrentComponentState<S>();
    const previousState = reconciler.getPreviousComponentState<S>();

    if (hookIndex in currentState) {
      reconciler.currentHookIndex++;
      return [currentState[hookIndex], stateSetter];
    }

    if (reconciler.isInitialRender || !(hookIndex in previousState)) {
      const value = initialState instanceof Function ? initialState() : initialState;

      if (value instanceof Promise) {
        const asyncResolver = async (): Promise<void> => {
          currentState[hookIndex] = await value;
          reconciler.currentHookIndex = 0;
        };
        throw asyncResolver();
      }

      currentState[hookIndex] = value;
    } else {
      currentState[hookIndex] = previousState[hookIndex];
    }

    function stateSetter(valueOrFunction: SetStateAction<S>): void {
      if (reconciler.isRendering) {
        throw new Error('Cannot call setState while rendering.');
      }

      currentState[hookIndex] =
        valueOrFunction instanceof Function
          ? valueOrFunction(currentState[hookIndex])
          : valueOrFunction;
    }

    reconciler.currentHookIndex++;

    return [currentState[hookIndex], stateSetter];
  }
  return useState;
}
