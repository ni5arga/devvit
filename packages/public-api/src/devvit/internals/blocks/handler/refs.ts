import { _activeRenderContext } from './BlocksHandler.js';
import type { HookRef } from './types.js';

/**
 * This function is mostly used for tests.  It's a way to capture the hook id without changing code too much
 *
 * Example:
 *
 * ```tsx
 * const ref: HookRef = {};
 * let [count, setCount] = captureHookRef(useState(0), ref);
 * ```
 *
 * @param value
 * @param ref
 * @returns
 */
export function captureHookRef<T>(value: T, ref: HookRef): T {
  if (!_activeRenderContext) {
    throw new Error('No active render context');
  }

  if (typeof value === 'function') {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (value as any)['captureHookRef'] = () => {
      if (!_activeRenderContext) {
        throw new Error('No active render context');
      }
      const id = _activeRenderContext._prevHookId;
      ref.id = id;
    };
  } else {
    const id = _activeRenderContext._prevHookId;
    if (!id) {
      throw new Error('No hook id found');
    }
    ref.id = id;
  }

  return value;
}
