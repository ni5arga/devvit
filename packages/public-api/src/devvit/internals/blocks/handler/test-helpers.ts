import type { UIEvent, UIRequest } from '@devvit/protos';
import { Header } from '@devvit/shared-types/Header.js';
import type { JSONValue } from '@devvit/shared-types/json.js';
import { _latestBlocksHandler } from './BlocksHandler.js';
import type { HookRef } from './types.js';

export const findHookId = (ref: HookRef): string => {
  return ref.id!;
};

export const findHookState = (ref: HookRef): JSONValue => {
  return _latestBlocksHandler!._latestRenderContext!._state[ref.id!];
};

export const EmptyRequest: UIRequest = { events: [] };

export const generatePressRequest = (ref: HookRef): UIRequest => {
  const event: UIEvent = {
    hook: ref.id!,
    userAction: {
      actionId: ref.id!,
    },
  };
  return {
    state: _latestBlocksHandler?._latestRenderContext?._state ?? {},
    events: [event],
  };
};

export const generateTimerRequest = (ref: HookRef): UIRequest => {
  const event: UIEvent = {
    hook: ref.id!,
    timer: {},
  };
  return {
    state: _latestBlocksHandler?._latestRenderContext?._state ?? {},
    events: [event],
  };
};

export const mockMetadata = {
  [Header.AppUser]: {
    values: ['t2_appuser'],
  },
  [Header.Subreddit]: {
    values: ['t5_devvit'],
  },
  [Header.User]: {
    values: ['t2_user'],
  },
};
