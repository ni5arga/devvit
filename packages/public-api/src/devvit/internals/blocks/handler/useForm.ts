import type { JSONValue } from '@devvit/shared-types/json.js';
import { getFormValues } from '../../../../apis/ui/helpers/getFormValues.js';
import type { Form, FormDefinition, FormFunction, FormKey, FormValues } from '../../../../index.js';
import { registerHook } from './BlocksHandler.js';
import type { RenderContext } from './RenderContext.js';
import type { EventHandler, Hook, HookRef } from './types.js';

class UseFormHook implements Hook, FormDefinition {
  hookId: string;
  state: JSONValue = null;
  onUIEvent?: EventHandler | undefined;
  onStateLoaded?: (() => void) | undefined;
  form: Form | FormFunction;
  onSubmit: (values: FormValues) => void | Promise<void>;

  constructor(
    params: { hookId: string },
    form: Form | FormFunction,
    onSubmit: (values: FormValues) => void | Promise<void>
  ) {
    this.hookId = params.hookId;
    this.form = form;
    this.onSubmit = onSubmit;
    this.onUIEvent = async (event) => {
      if (event.formSubmitted) {
        await onSubmit(getFormValues(event.formSubmitted.results));
      }
    };
  }
}

export function useForm(
  form: Form | FormFunction,
  onSubmit: (values: FormValues) => void | Promise<void>
): FormKey {
  const hook = registerHook(
    { namespace: 'useForm' },
    (params) => new UseFormHook(params, form, onSubmit)
  );
  return hookRefToFormKey({ id: hook.hookId });
}

export function hookRefToFormKey(hookRef: HookRef): FormKey {
  return `form.hook.${hookRef.id}.0`;
}

export function formKeyToHookRef(formKey: FormKey): HookRef {
  // extract the hook id from the form key with a regex
  return { id: formKey.match(/form\.hook\.(.+)\.0/)![1] };
}

export function getFormDefinition(renderContext: RenderContext, formKey: FormKey): FormValues {
  const hookId = formKeyToHookRef(formKey);
  return renderContext.getHook(hookId) as UseFormHook;
}
