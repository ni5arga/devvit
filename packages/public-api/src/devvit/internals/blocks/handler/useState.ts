import { EffectType } from '@devvit/protos';
import type { JSONValue } from '@devvit/shared-types/json.js';
import type {
  AsyncUseStateInitializer,
  SetStateAction,
  UseStateInitializer,
  UseStateResult,
} from '../../../../types/hooks.js';
import { registerHook } from './BlocksHandler.js';
import type { RenderContext } from './RenderContext.js';
import type { Hook, HookParams } from './types.js';
import { RenderInterruptError } from './types.js';

export function useState(initialState: UseStateInitializer<boolean>): UseStateResult<boolean>;
export function useState(initialState: UseStateInitializer<number>): UseStateResult<number>;
export function useState(initialState: UseStateInitializer<string>): UseStateResult<string>;
export function useState<S extends JSONValue>(
  initialState: UseStateInitializer<S>
): UseStateResult<S>;
export function useState<S extends JSONValue>(
  initialState: UseStateInitializer<S>
): UseStateResult<S> {
  const hook = registerHook({ namespace: 'useState' }, ({ changed }) => {
    const state = initialState instanceof Function ? initialState() : initialState;
    if (state instanceof Promise) {
      throw new Error('Cannot use async initializer with useState, use useAsyncState instead.');
    }
    const setter = (action: SetStateAction<S>): void => {
      hook.state = action instanceof Function ? action(hook.state) : action;
      changed();
    };
    return { state, setter };
  });
  return [hook.state, hook.setter];
}

class AsyncStateHook<S extends JSONValue> implements Hook {
  state: { value: S | null; loading: boolean } = { value: null, loading: false };
  #changed: () => void;
  #initializer: AsyncUseStateInitializer<S>;
  #hookId: string;

  constructor(initializer: AsyncUseStateInitializer<S>, params: HookParams) {
    this.#initializer = initializer;
    this.#hookId = params.hookId;
    this.#changed = params.changed;
  }

  async onUIEvent(): Promise<void> {
    if (this.state.value === null) {
      this.state.value = await this.#initializer();
      this.state.loading = false;
      this.#changed();
    }
  }

  setter(action: SetStateAction<S>): void {
    this.state.value = action instanceof Function ? action(this.state.value as S) : action;
    this.#changed();
  }

  onLoad(renderContext: RenderContext): void {
    if (this.state.value === null && !this.state.loading) {
      this.state.loading = true;
      this.#changed();
      renderContext.emitEffect(this.#hookId, {
        type: EffectType.EFFECT_SEND_EVENT,
        // This is on the main queue, because it blocks app load.
        sendEvent: { event: { asyncRequest: { requestId: this.#hookId }, hook: this.#hookId } },
      });
      throw new RenderInterruptError();
    }
  }
}

export function useAsyncState<S extends JSONValue>(
  initializer: AsyncUseStateInitializer<S>
): UseStateResult<S> {
  const hook = registerHook({ namespace: 'useAsyncState' }, (params) => {
    return new AsyncStateHook(initializer, params);
  });

  return [hook.state.value as S, hook.setter.bind(hook)];
}
