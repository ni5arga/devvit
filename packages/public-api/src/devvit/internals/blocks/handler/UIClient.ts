import type { Toast as ToastProto } from '@devvit/protos';
import { EffectType, Form, ToastAppearance } from '@devvit/protos';
import type { Data } from '../../../../types/data.js';
import type { FormKey } from '../../../../types/form.js';
import type { Toast } from '../../../../types/toast.js';
import type { UIClient as _UIClient } from '../../../../types/ui-client.js';
import { assertValidFormFields } from '../../../../apis/ui/helpers/assertValidFormFields.js';
import { transformFormFields } from '../../../../apis/ui/helpers/transformForm.js';
import type { Comment, Post, Subreddit, User } from '../../../../apis/reddit/models/index.js';
import type { RenderContext } from './RenderContext.js';
import { _activeRenderContext } from './BlocksHandler.js';
import { getFormDefinition } from './useForm.js';

export function useUI(): _UIClient {
  const renderContext = _activeRenderContext;
  if (!renderContext) {
    throw new Error('useUI can only be called from within the top level of a component.');
  }
  return new UIClient(renderContext);
}

export class UIClient implements _UIClient {
  #renderContext: RenderContext;

  constructor(renderContext: RenderContext) {
    this.#renderContext = renderContext;
  }

  showForm(formKey: FormKey, data?: Data | undefined): void {
    const formDefinition = getFormDefinition(this.#renderContext, formKey);

    if (!formDefinition) {
      throw new Error('Form does not exist. Make sure you have added it using useForm.');
    }

    const formData =
      formDefinition.form instanceof Function
        ? formDefinition.form(data ?? {})
        : formDefinition.form;

    const form = Form.fromPartial({
      id: formKey,
      title: formData.title,
      acceptLabel: formData.acceptLabel,
      cancelLabel: formData.cancelLabel,
      shortDescription: formData.description,
    });

    assertValidFormFields(formData.fields);
    form.fields = transformFormFields(formData.fields);

    this.#renderContext.emitEffect(formKey, {
      type: EffectType.EFFECT_SHOW_FORM,
      showForm: {
        form,
      },
    });
  }

  showToast(text: string): void;
  showToast(toast: Toast): void;
  showToast(textOrToast: string | Toast): void {
    let toast: ToastProto;

    if (textOrToast instanceof Object) {
      toast = {
        text: textOrToast.text,
        appearance:
          textOrToast.appearance === 'success' ? ToastAppearance.SUCCESS : ToastAppearance.NEUTRAL,
      };
    } else {
      toast = {
        text: textOrToast,
      };
    }

    this.#renderContext.emitEffect(textOrToast.toString(), {
      type: EffectType.EFFECT_SHOW_TOAST,
      showToast: {
        toast,
      },
    });
  }

  navigateTo(url: string): void;
  navigateTo(subreddit: Subreddit): void;
  navigateTo(post: Post): void;
  navigateTo(comment: Comment): void;
  navigateTo(user: User): void;
  navigateTo(thingOrUrl: string | Subreddit | Post | Comment | User): void {
    let url: string;

    if (typeof thingOrUrl === 'string') {
      // Validate URL
      url = new URL(thingOrUrl).toString();
    } else {
      url = new URL(thingOrUrl.permalink, 'https://www.reddit.com').toString();
    }
    this.#renderContext.emitEffect(url, {
      type: EffectType.EFFECT_NAVIGATE_TO_URL,
      navigateToUrl: {
        url,
      },
    });
  }
}
