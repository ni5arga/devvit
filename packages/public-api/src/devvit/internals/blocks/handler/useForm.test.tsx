/** @jsx Devvit.createElement */
/** @jsxFrag Devvit.Fragment */

import { describe, expect, test } from 'vitest';
import { hookRefToFormKey, useForm } from './useForm.js';
import { useState } from './useState.js';
import { useUI } from './UIClient.js';
import { Devvit } from '../../../Devvit.js';
import { EffectType, UIEvent, FormFieldType } from '@devvit/protos';
import { BlocksHandler } from './BlocksHandler.js';
import { EmptyRequest, generatePressRequest, mockMetadata } from './test-helpers.js';
import { captureHookRef } from './refs.js';
import type { HookRef } from './types.js';

const buttonRef: HookRef = {};
const formRef: HookRef = {};
let submitCount = 0;

const App = (): JSX.Element => {
  const ui = useUI();
  const form = captureHookRef(
    useForm(
      {
        title: 'Test Form',
        description: 'This is a test form',
        acceptLabel: 'Yup',
        cancelLabel: 'Nope',
        fields: [
          {
            type: 'string',
            name: 'name',
            label: 'Name',
          },
        ],
      },
      () => {
        submitCount++;
      }
    ),
    formRef
  );
  return (
    <button
      onPress={captureHookRef(() => {
        ui.showForm(form);
      }, buttonRef)}
    >
      Open Form
    </button>
  );
};

describe('useForm', () => {
  test('emits an effect on render', async () => {
    const handler = new BlocksHandler(App);
    await handler.handle(EmptyRequest, mockMetadata);
    const response = await handler.handle(generatePressRequest(buttonRef), mockMetadata);
    expect(response).toMatchSnapshot();
    expect(response.effects).toHaveLength(1);
    expect(response.effects[0].type).toBe(EffectType.EFFECT_SHOW_FORM);
  });

  test('calls onSubmit when a form is submitted', async () => {
    submitCount = 0;
    const handler = new BlocksHandler(App);
    await handler.handle(EmptyRequest, mockMetadata);
    const response = await handler.handle(generatePressRequest(buttonRef), mockMetadata);
    const event: UIEvent = UIEvent.fromPartial({
      formSubmitted: {
        formId: hookRefToFormKey(formRef),
        results: {
          name: {
            stringValue: 'Genghis Corgi',
          },
        },
      },
      hook: formRef.id,
    });
    const req = { events: [event], state: response.state };
    await handler.handle(req, mockMetadata);
    expect(submitCount).toBe(1);
  });

  test('test rendering FormSubmittedEvent', async () => {
    const App = (): JSX.Element => {
      const ui = useUI();
      const [name, setName] = useState('');
      const form = captureHookRef(
        useForm(
          {
            title: 'Test Form',
            description: 'This is a test form',
            fields: [
              {
                type: 'string',
                name: 'name2',
                label: 'Name',
              },
            ],
          },
          ({ name }) => setName(name)
        ),
        formRef
      );
      return (
        <vstack>
          <button
            onPress={captureHookRef(() => {
              ui.showForm(form);
            }, buttonRef)}
          >
            Open Form
          </button>
          <text size="large">{name}</text>;
        </vstack>
      );
    };

    const handler = new BlocksHandler(App);
    await handler.handle(EmptyRequest, mockMetadata);
    const pressRsp = await handler.handle(generatePressRequest(buttonRef), mockMetadata);
    const event: UIEvent = {
      formSubmitted: {
        formId: hookRefToFormKey(formRef),
        results: {
          name: {
            stringValue: 'Genghis Corgi',
            fieldType: FormFieldType.STRING,
            numberValue: undefined,
            boolValue: undefined,
          },
        },
      },
      hook: formRef.id,
    };
    const submitReq = { events: [event], state: pressRsp.state };
    const submitRsp = await handler.handle(submitReq, mockMetadata);
    expect(submitRsp).toMatchSnapshot();
  });
});
