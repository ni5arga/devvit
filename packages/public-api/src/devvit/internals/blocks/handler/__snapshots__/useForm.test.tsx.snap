// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useForm > emits an effect on render 1`] = `
{
  "blocks": {
    "actions": [],
    "config": {
      "rootConfig": {
        "children": [
          {
            "actions": [
              {
                "data": {},
                "id": "App.button.onPress",
                "type": 0,
              },
            ],
            "config": {
              "buttonConfig": {
                "buttonAppearance": undefined,
                "buttonSize": undefined,
                "disabled": undefined,
                "icon": undefined,
                "text": "Open Form",
                "textColor": undefined,
                "textColors": undefined,
              },
            },
            "id": undefined,
            "key": undefined,
            "size": undefined,
            "sizes": undefined,
            "type": 3,
          },
        ],
        "height": 320,
      },
    },
    "id": undefined,
    "key": undefined,
    "size": undefined,
    "sizes": undefined,
    "type": 0,
  },
  "effects": [
    {
      "showForm": {
        "form": {
          "acceptLabel": "Yup",
          "cancelLabel": "Nope",
          "fields": [
            {
              "defaultValue": {
                "boolValue": undefined,
                "fieldType": 0,
                "groupValue": undefined,
                "isSecret": undefined,
                "listValue": undefined,
                "numberValue": undefined,
                "selectionValue": undefined,
                "stringValue": undefined,
              },
              "disabled": undefined,
              "fieldConfig": {
                "booleanConfig": undefined,
                "groupConfig": undefined,
                "listConfig": undefined,
                "numberConfig": undefined,
                "paragraphConfig": undefined,
                "selectionConfig": undefined,
                "stringConfig": {
                  "maxLength": undefined,
                  "minLength": undefined,
                  "placeholder": undefined,
                },
              },
              "fieldId": "name",
              "fieldType": 0,
              "helpText": undefined,
              "isSecret": undefined,
              "label": "Name",
              "required": undefined,
            },
          ],
          "id": "form.hook.App.useForm-0.0",
          "shortDescription": "This is a test form",
          "title": "Test Form",
        },
      },
      "type": 3,
    },
  ],
  "state": {
    "__generation": 1,
  },
}
`;

exports[`useForm > test rendering FormSubmittedEvent 1`] = `
{
  "blocks": {
    "actions": [],
    "config": {
      "rootConfig": {
        "children": [
          {
            "actions": [],
            "config": {
              "stackConfig": {
                "alignment": undefined,
                "backgroundColor": undefined,
                "backgroundColors": undefined,
                "border": undefined,
                "children": [
                  {
                    "actions": [
                      {
                        "data": {},
                        "id": "App2.vstack.button-0.onPress",
                        "type": 0,
                      },
                    ],
                    "config": {
                      "buttonConfig": {
                        "buttonAppearance": undefined,
                        "buttonSize": undefined,
                        "disabled": undefined,
                        "icon": undefined,
                        "text": "Open Form",
                        "textColor": undefined,
                        "textColors": undefined,
                      },
                    },
                    "id": undefined,
                    "key": "0",
                    "size": undefined,
                    "sizes": undefined,
                    "type": 3,
                  },
                  {
                    "actions": [],
                    "config": {
                      "textConfig": {
                        "alignment": undefined,
                        "color": undefined,
                        "colors": undefined,
                        "outline": undefined,
                        "overflow": 1,
                        "selectable": undefined,
                        "size": 3,
                        "style": undefined,
                        "text": "Genghis Corgi",
                        "weight": undefined,
                        "wrap": undefined,
                      },
                    },
                    "id": undefined,
                    "key": "1",
                    "size": undefined,
                    "sizes": undefined,
                    "type": 2,
                  },
                ],
                "cornerRadius": undefined,
                "direction": 1,
                "gap": undefined,
                "padding": undefined,
                "reverse": undefined,
              },
            },
            "id": undefined,
            "key": undefined,
            "size": undefined,
            "sizes": undefined,
            "type": 1,
          },
        ],
        "height": 320,
      },
    },
    "id": undefined,
    "key": undefined,
    "size": undefined,
    "sizes": undefined,
    "type": 0,
  },
  "effects": [],
  "state": {
    "App2.useState-0": "Genghis Corgi",
    "__generation": 2,
  },
}
`;
